src/Solve.cpp

```diff
@@ -345,31 +344,160 @@ class SolveExpression : public IRMutator {
+        const Sub *sub_b = b.as<Sub>();
+        const Mul *mul_a = a.as<Mul>();
+        const Mul *mul_b = b.as<Mul>();
+        const T *t_a = a.as<T>();
+        const T *t_b = b.as<T>();
+
+        expr = Expr();
+
+        if (a_uses_var && !b_uses_var) {
+            if (t_a) {
+                // op(op(f(x), a), b) -> op(f(x), op(a, b))
+                expr = mutate(T::make(t_a->a, T::make(t_a->b, b)));
+            }
+        } else if (a_uses_var && b_uses_var) {
+            if (equal(a, b)) {
+                // op(f(x), f(x)) -> f(x)
+                expr = a;
+            } else if (t_a) {
+                // op(op(f(x), a), g(x)) -> op(op(f(x), g(x)), a)
+                expr = mutate(T::make(T::make(t_a->a, b), t_a->b));
+            } else if (t_b) {
+                // op(f(x), op(g(x), a)) -> op(op(f(x), g(x)), a)
+                expr = mutate(T::make(T::make(a, t_b->a), t_b->b));
+            } else if (add_a && add_b && equal(add_a->a, add_b->a)) {
+                // op(f(x) + a, f(x) + b) -> f(x) + op(a, b)
+                expr = mutate(add_a->a + T::make(add_a->b, add_b->b));
+            } else if (add_a && add_b && equal(add_a->b, add_b->b)) {
+                // op(f(x) + a, g(x) + a) -> op(f(x), g(x)) + a;
+                expr = mutate(T::make(add_a->a, add_b->a)) + add_a->b;
+            } else if (add_a && equal(add_a->a, b)) {
+                // op(f(x) + a, f(x)) -> f(x) + op(a, 0)
+                expr = mutate(b + T::make(add_a->b, 0));
+            } else if (add_b && equal(add_b->a, a)) {
+                // op(f(x), f(x) + a) -> f(x) + op(a, 0)
+                expr = mutate(a + T::make(add_b->b, 0));
+            } else if (sub_a && sub_b && equal(sub_a->a, sub_b->a)) {
+                // op(f(x) - a, f(x) - b) -> f(x) - op(a, b)
+                expr = mutate(sub_a->a - T::make(sub_a->b, sub_b->b));
+            } else if (sub_a && sub_b && equal(sub_a->b, sub_b->b)) {
+                // op(f(x) - a, g(x) - a) -> op(f(x), g(x)) - a;
+                expr = mutate(T::make(sub_a->a, sub_b->a)) - sub_a->b;
+            } else if (sub_a && equal(sub_a->a, b)) {
+                // op(f(x) - a, f(x)) -> f(x) - op(a, 0)
+                expr = mutate(b - T::make(sub_a->b, 0));
+            } else if (sub_b && equal(sub_b->a, a)) {
+                // op(f(x), f(x) - a) -> f(x) - op(a, 0)
+                expr = mutate(a - T::make(sub_b->b, 0));
+            } else if (mul_a && mul_b && equal(mul_a->a, mul_b->a)) {
+                // op(f(x)*a, f(x)*b) -> f(x)*op(a, b)
+                expr = mutate(mul_a->a * T::make(mul_a->b, mul_b->b));
+            } else if (mul_a && mul_b && equal(mul_a->b, mul_b->b)) {
+                // op(f(x)*a, g(x)*a) -> op(f(x), g(x))*a;
+                expr = mutate(T::make(mul_a->a, mul_b->a)) * mul_a->b;
+            } else if (mul_a && equal(mul_a->a, b)) {
+                // op(f(x)*a, f(x)) -> f(x) * op(a, 1)
+                expr = mutate(b * T::make(mul_a->b, 1));
```

Comment:
Wait... some of these multiply rules are wrong:

max(a*b, a*c) != a*max(b, c) if a is negative